core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@lindong28 : Thanks for the reply. I assume that when you do rolling bounce, the process is to (a) controlled shut down a broker; (b) restart the broker; (c) wait for all replicas to be in sync; (d) move to the next broker. Then, I am trying to understand what component is improved by a factor of 10. Based on your previous answer, it seems that most of the speedup is in step (c)? My understanding is that if the request queue builds up in the broker, it delays some partitions to become the leader and therefore delays the followers to be in sync. Is that correct? My next question is why would the request queue build up. Is it because there is not enough request handler threads or is it because the handling of a LeaderAndIsrRequest is expensive? Note that since each ZK write includes a disk flush and an extra RPC, I was expecting that the handling of each LeaderAndIsrRequest to be cheaper.

Now, assuming that batching makes a big difference. Another potential way to implement that is to leave the logic in controller/state-machines unchanged, but change how requests are sent in RequestSendThread instead. After RequestSendThread completes the send of a request to a broker, multiple requests may have been accumulated by the controller. We can potentially dequeue a batch of those request, collapse them into a single request and send it to the broker. That might be simpler than changing controller logic.